camel/memories/blocks/graphdb_block.py

```diff
@@ -0,0 +1,224 @@
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import logging
+from enum import Enum
+from typing import List, Optional
+
+from camel.embeddings import BaseEmbedding, OpenAIEmbedding
+from camel.memories.base import MemoryBlock
+from camel.memories.records import ContextRecord, MemoryRecord
+from camel.storages.graph_storages import BaseGraphStorage, Neo4jGraph
+from camel.types import OpenAIBackendRole, RoleType
+
+logger = logging.getLogger(__name__)
+
+
+class GraphDBBlock(MemoryBlock):
+    r"""An implementation of the MemoryBlock abstract base class for
+    maintaining and retrieving information using a graph database.
+
+    Args:
+        storage (Optional[BaseGraphStorage], optional): The storage mechanism
+            for the graph database. Defaults to a specific implementation
+            if not provided. (default: None)
+        embedding (Optional[BaseEmbedding], optional): Embedding mechanism
+            toconvert text into vector representations. Defaults to
+            OpenAIEmbedding if not provided. (default: None)
+    """
+
+    def __init__(
+        self,
+        storage: Optional[BaseGraphStorage] = None,
+        embedding: Optional[BaseEmbedding] = None,
+        url: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        database: Optional[str] = "neo4j",
+    ) -> None:
+        self.embedding = embedding or OpenAIEmbedding()
+        self.storage = storage or Neo4jGraph(url, username, password, database)
+
+    def retrieve(
+        self,
+        query_embedding: Optional[List[float]] = None,
+        limit: int = 25,
+    ) -> List[ContextRecord]:
+        r"""Retrieves records from the graph database based on query embedding.
+
+        Args:
+            query_embedding (Optional[List[float]]): The embedding vector
+                for similarity search.
+            limit (int): The maximum number of records to retrieve.
+                Defaults to 10.
+            order_by (str): The property to order by in the default query.
+                Defaults to "timestamp".
+            desc (bool): Whether to order in descending order.
+                Defaults to True.
+
+        Returns:
+            List[ContextRecord]: A list of context records retrieved
+                from the graph database.
+        """
+        try:
+            if query_embedding:
+                # Perform a similarity search based on the provided embedding
+                results = self.storage.query(
+                    """
+                    MATCH (n:Message)
+                    WHERE exists(n.embedding)
+                    WITH n, gds.similarity.cosine(
```

Comment:
we can use `db.index.vector.queryNodes` instead, since gds need additional dependency